### DISCUSSION TITLE
Reading Zygo MetroPro file

### DISCUSSION THREAD

**Discussion by nombac** [ContactEngineering/SurfaceTopography#332] [General] [answered]
I began using SurfaceTopography to read Zygo MetroPro files. Upon comparing an image generated by SurfaceTopography (along with matplotlib shown below) to one produced by Zygo's Mx software, it appears that the x and y coordinates of the height array have been swapped, as shown in the attachment. What could be the cause of the issue?

```
import SurfaceTopography as st
import matplotlib.pyplot as plt
import numpy as np
import pint

ureg = pint.UnitRegistry()

def display_surf_file(filename):
    # Load topography from file
    topography = st.read_topography(filename)

    # Extract height information
    heights = topography.heights()  # Transpose the heights matrix
    #heights = np.transpose(topography.heights())  # Transpose the heights matrix

    # Convert the physical sizes to desired unit (um)
    # Convert the heights to nanometers
    unit_from_file = ureg.Unit(topography.unit)
    conversion_factor = (1 * unit_from_file).to('um').magnitude
    physical_sizes_in_um = tuple(p * conversion_factor for p in topography.physical_sizes)
    conversion_factor_to_nm = (1 * unit_from_file).to('nm').magnitude
    heights_in_nm = heights * conversion_factor_to_nm

    # Create an evenly spaced range based on the physical sizes in um and pixel resolution
    x = np.linspace(0, physical_sizes_in_um[0], heights_in_nm.shape[0])
    y = np.linspace(0, physical_sizes_in_um[1], heights_in_nm.shape[1])

    # Plot the data using extent to adjust the axes
    vmin = -20
    vmax =  20
    cmap = 'viridis'
    # Plot the data using extent to adjust the axes
    plt.imshow(heights_in_nm, cmap='viridis', extent=[x.min(), x.max(), y.max(), y.min()], vmin=vmin, vmax=vmax)

    ticks = np.arange(0, max(physical_sizes_in_um), 20)  # Example: increments of 25 um
    plt.xticks(ticks)
    plt.yticks(ticks)

    plt.colorbar(label='Height (nm)')  # Updated the label to nanometers
    plt.title(filename)  # Use the filename as the title
    plt.xlabel(f'X (um)')
    plt.ylabel(f'Y (um)')
    plt.show()

if __name__ == "__main__":
    filename = 'NS030L_003_processed.dat'
    display_surf_file(filename)
```

[20231011_st.pdf](https://github.com/ContactEngineering/SurfaceTopography/files/12862720/20231011_st.pdf)

**Comment by pastewka** [comment, marked as answer]
This is the data layout, thanks for notifying us. I will swap it, but note that this does not affect any of the statistical properties that you may compute using the code.

**Reply by nombac** [reply]
Thank you for your answer. (I'd appreciate it if you could let me know when the swap is done.)

**Reply by pastewka** [reply]
Hi @nombac - the data map should be swapped starting version 1.11.1. Can you check whether this works for you?

**Comment by nombac** [comment]
I've confirmed that version 1.11.1 works for me.
Thank you very much for working on this.

**Reply by pastewka** [reply]
Perfect, thanks for confirming (and for reporting the issue)!

**Comment by nombac** [comment]
It seems that Gwyddion format files (.gwy) also need the swap (I'm using 1.11.3).

**Reply by pastewka** [reply]
Okay, thanks for notifying me. Will fix this.

**Comment by pastewka** [comment]
@nombac Is this still an issue for .gwy files. It looks okay if I load and plot them, but it is correct that the origin in the plot below is in different locations in Gwyddion and SurfaceTopography

<img width="1340" alt="image" src="https://github.com/ContactEngineering/SurfaceTopography/assets/2530391/d0c0dea5-e690-442f-b65e-dc003efd8256">

**Reply by pastewka** [reply]
I guess you want the y-axis swapped (but I should then also adjust the plot function to invert the y-axis).

**Reply by nombac** [reply]
Just to clarify the orientation issue with the heights data from different formats, I've attached diagrams comparing the three cases:

- metropro format: topo.heights()
- Gwyddion native format: topo.heights()
- Gwyddion native format after rotation: np.rot90(topo.heights())

It seems that topo.heights() from the Gwyddion native format needs a np.rot90 rotation to match the metropro format orientation, which is consistent with how these files are displayed in Gwyddion.

![20240417_SurfaceTopography 001](https://github.com/ContactEngineering/SurfaceTopography/assets/38897846/2f199d60-6f73-48fa-9257-6fa5c9ff3a4d)

**Reply by nombac** [reply]
Here are the links to the files I used in the above example:
[test.gwy](https://www.dropbox.com/scl/fi/ssk39frrbyegdjaisbs77/test.gwy?rlkey=96nrivr5zy9gocelarzjhus0g&dl=0)
[test.dat](https://www.dropbox.com/scl/fi/60vfyzybsjgctvd33ng4l/test.dat?rlkey=krsrfnrgypo7b7tybwd1sg14l&dl=0)

**Comment by pastewka** [comment]
@nombac  - This was fixed in v1.13.0. Can you confirm that this is indeed working for you now?

**Reply by pastewka** [reply]
I am referring to the Gwyddion flip

**Comment by nombac** [comment]
Sorry for the late reply.
The issue still appears to persist in version 1.16.2 as attached.

**Reply by pastewka** [reply]
Okay, will look into this!
